src/lexer/init.lua

```diff
@@ -263,6 +264,51 @@ function lexer.navigator()
 		end)
 	end
 
+	function nav:HotswapSource(newSource)
+		local oldSource = self.Source or ""
+		self.Source = newSource
+
+		local minimumLength, maximumLength = 0, math.min(#oldSource, #newSource)
+		while minimumLength < maximumLength do
+			local mid = (minimumLength + maximumLength + 1) // 2
+
+			if
+				oldSource:byte(mid) == newSource:byte(mid) -- cheap check of last character
+				and oldSource:sub(1, mid - 1) == newSource:sub(1, mid - 1) -- expensive check of all previous characters
+			then
+				minimumLength = mid
+			else
+				maximumLength = mid - 1
+			end
+		end
+
+		local lastSharedTokenIndex = 0
+		local sourceIndex = 0
+
+		-- find position of the last common token that is cached
+		for tokenIndex, cachedToken in self.TokenCache do
+			local sourceLength = #cachedToken[2]
+
+			if sourceIndex + sourceLength < minimumLength then
+				lastSharedTokenIndex = tokenIndex
+				sourceIndex += sourceLength
+			else
+				break
+			end
+		end
+
+		self._RealIndex = math.min(self._RealIndex, lastSharedTokenIndex)
```

Comment:
All the indexes past this new `RealIndex` in the token cache should to be set to nil as they are now outdated.